### PR TITLE
Ext proc http functionality support

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
Risk Level: low
Testing: n/a
Docs Changes: n/a
Release Notes: inline
Fixes:

Description:
This is to address the issue: https://github.com/envoyproxy/envoy/issues/35488, i.e, integrate the ext_proc HTTP client to ext_proc filter.  With this PR, the basic functionalities to have Envoy ext_proc filter talk to a HTTP server using HTTP messages are accomplished.

This is the follow up of PR: https://github.com/envoyproxy/envoy/pull/35676

### CODE REVIEW

**Comment by yanjunxiang-google** [comment]
This PR adds the support of sending ext_proc call out messages to the HTTP side stream server.  The proto messages: ProcessingRequest/ProcessingResponse are transcoded into JSON text and encoded in the body of a HTTP message.

The processing flow works this way:

1) When Envoy needs to send call out messages to the side stream server, either header, body , or trailer, Envoy construct a ProcessingRequest proto message.
2) This ProcessingRequest message is transcoded into a JSON text. This JSON text is then added as a body to a HTTP "POST" message. This HTTP message is sent to the side stream HTTP server.
3) The HTTP server receives the "POST" message. Transcoding the body, i.e, a JSON text, into a ProcessingRequest message.
4) The HTTP server then performs normal header/body/trailer mutation as existing, and put them into a ProcessingResponse proto message.
5)  The HTTP server then transcode the ProcessingResponse into JSON text, and sends back a 200 response with this JSON text as body.
6) After receives the 200 response, Envoy then convert the body from JSON text back to a ProcessingResponse proto message. Then continue processing this ProcessingResponsel.

**Comment by htuch** [comment]
@yanjunxiang-google thoughts on how the receiving server correlates HTTP requests? There are two things here and I think we need to do both, to avoid making users solve a distributed system coordination problem:
1. Use HTTP session affinity. I'm not sure if we have a strong session affinity mechanism here that doesn't rely on cookies (and hence doesn't work for ext_proc initiated HTTP) though.
2. Attach a UUID (e.g. the x-envoy-request-id) as a distinguished field to the JSON to allow the server to correlate.

**Comment by tyxia** [comment]
> @yanjunxiang-google thoughts on how the receiving server correlates HTTP requests for a given dataplane streams? There are two things here and I think we need to do both, to avoid making users solve a distributed system coordination problem:
> 
> 1. Use HTTP session affinity. I'm not sure if we have a strong session affinity mechanism here that doesn't rely on cookies (and hence doesn't work for ext_proc initiated HTTP) though.
> 2. Attach a UUID (e.g. the x-envoy-request-id) as a distinguished field to the JSON to allow the server to correlate.

Here is related design and internal discussion https://docs.google.com/document/d/1UDPm1J4txb2P2oCnl98qClD2FMZWu3SdnRnoHJZxGBU/edit?disco=AAABM_WD00c
Regarding 1, Should we suggest gRPC as recommend protocol for strong affinity.

**Comment by yanjunxiang-google** [comment]
> @yanjunxiang-google thoughts on how the receiving server correlates HTTP requests for a given dataplane streams? There are two things here and I think we need to do both, to avoid making users solve a distributed system coordination problem:
> 
> 1. Use HTTP session affinity. I'm not sure if we have a strong session affinity mechanism here that doesn't rely on cookies (and hence doesn't work for ext_proc initiated HTTP) though.
> 2. Attach a UUID (e.g. the x-envoy-request-id) as a distinguished field to the JSON to allow the server to correlate.


Good suggestions!

I put this as the step 7 of the issue: https://github.com/envoyproxy/envoy/issues/35488, and plan to add this in my next PR.
BTW, we will only remove the [#not-implemented-hide:] from the API :
https://github.com/envoyproxy/envoy/blob/84c8da7540e4e5064f6a6dffc66825668f6220e4/api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto#L134 after all the steps listed in the issue https://github.com/envoyproxy/envoy/issues/35488 completed.

Please continue review this PR and let me know the comments.

**Comment by yanjunxiang-google** [comment]
> > @yanjunxiang-google thoughts on how the receiving server correlates HTTP requests for a given dataplane streams? There are two things here and I think we need to do both, to avoid making users solve a distributed system coordination problem:
> > 
> > 1. Use HTTP session affinity. I'm not sure if we have a strong session affinity mechanism here that doesn't rely on cookies (and hence doesn't work for ext_proc initiated HTTP) though.
> > 2. Attach a UUID (e.g. the x-envoy-request-id) as a distinguished field to the JSON to allow the server to correlate.
> 
> Here is related design and internal discussion https://docs.google.com/document/d/1UDPm1J4txb2P2oCnl98qClD2FMZWu3SdnRnoHJZxGBU/edit?disco=AAABM_WD00c
> 
> @htuch Regarding point 1 above, Should we consider suggesting gRPC as the recommend protocol for strong affinity.

Thanks for the info!

To add the support of 2) is straightforward.  Can we make decision on whether we need to support 1) for ext_proc HTTP?

**Comment by htuch** [comment]
I think we should recommend gRPC for affinity, but I know inevitably we will end up with folks who need to solve this problem the moment we support > 1 event at a time on the HTTP variant. I think it's pretty rare to solve this via some distributed state sharing mechanism on the service callout, so you will need affinity.

At a minimum, please do (2) and document as appropriately the limitations for users of the HTTP variant, leave a bug open for (1), thanks.

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc [hunk 1]
```diff
@@ -1,20 +1,81 @@
...
+  if (!req_in_json.ok()) {
+    ENVOY_LOG(error,
+              "The ProcessingRequest proto message can not be encoded into a JSON string: {}",
+              req_in_json.status().ToString());
+    onError();
+    return;
+  }
+
+  const std::string cluster = config_.http_service().http_service().http_uri().cluster();
+  const std::string url = config_.http_service().http_service().http_uri().uri();
```
Comment (yanavlasov) [inline source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:28]: Is there any validation to make sure the URI is valid and produces valid host and path?

**Review by yanavlasov — COMMENTED** [review]
/wait-any

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc (hunk 1 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:28]: This is Envoy filter configuration from trusted domain. I think we do not need validation.  If this turns out to be a bad configuration, the outcome will be no response received by ext_proc filter, and timeout happens. This will trigger customer to go back to fix the configuration.

**Comment by yanjunxiang-google** [comment]
> I think we should recommend gRPC for affinity, but I know inevitably we will end up with folks who need to solve this problem the moment we support > 1 event at a time on the HTTP variant. I think it's pretty rare to solve this via some distributed state sharing mechanism on the service callout, so you will need affinity.
> 
> At a minimum, please do (2) and document as appropriately the limitations for users of the HTTP variant, leave a bug open for (1), thanks.

The x-request-id is added in the change now.

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc (hunk 1 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc]: @yanavlasov I checked a couple of other Envoy call out extensions which use http_uri service, like ext_authz, no configuration validation there either. So, let's keep it this way.  Please let me know if you have concerns.

**Comment by tyxia** [comment]
@yanjunxiang-google  Could you take a look at CI failures? The asan failures look legit.

/wait

**Comment by yanjunxiang-google** [comment]
@tyxia I am scaling back to only support forwarding headers in this PR, and leaving body and trailers forwarding as future tasks.

**Comment by yanjunxiang-google** [comment]
@stevenzzzz @tyxia @mattklein123 @yanavlasov  Kind Ping!

File: source/extensions/filters/http/ext_proc/http_client/client_base.h [hunk 2]
```diff
@@ -22,11 +25,10 @@ class RequestCallbacks {
 class ClientBase {
```
Comment (yanavlasov) [inline source/extensions/filters/http/ext_proc/http_client/client_base.h:25]: Is ClientBase common to both gRPC and HTTP clients?

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 3]
```diff
@@ -328,6 +352,52 @@ void Filter::setEncoderFilterCallbacks(Http::StreamEncoderFilterCallbacks& callb
   watermark_callbacks_.setEncoderFilterCallbacks(&callbacks);
 }
 
+void Filter::sendRequest(ProcessingRequest&& req, bool end_stream) {
+  if (config_->grpcService().has_value()) {
```
Comment (yanavlasov) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:356]: It looks like you trying to replicate C++ polymorphism here. Is ti possible to have one interface common to both gRPC and HTTP, so the `if` statement is not required?

**Review by yanavlasov — COMMENTED** [review]
/wait-any

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto [hunk 4]
```diff
@@ -126,11 +126,31 @@ message ExternalProcessor {
...
+  // sending request or response headers to the side stream server.
+  //
+  // With this configuration, Envoy behavior:
+  // 1. The headers are first put in a proto message
+  // :ref:`ProcessingRequest <envoy_v3_api_msg_service.ext_proc.v3.ProcessingRequest>`.
+  // 2. This proto message is then transcoded into a JSON text.
+  // 3. Envoy then sends a HTTP POST message with content-type as "application/json",
+  // and this JSON text as body to the side stream server.
+  //
+  // After receives this HTTP request message, the side stream server behavior:
```
Comment (tyxia) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:146]: nit: ext_proc server or external processor to be consistent with this file

But this is up to you

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto [hunk 5]
```diff
@@ -126,11 +126,31 @@ message ExternalProcessor {
...
+  // 1. It converts the body, which is a JSON string, into a ``ProcessingRequest`` proto message.
+  // 2. It performs whatever header mutation on the headers in this ``ProcessingRequest``.
```
Comment (tyxia) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:148]: These two lines number are duplicated and they can be merged into 1:
It can convert the body, which is a JSON string, into a ``ProcessingRequest`` proto message to examine the headers.

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc [hunk 6]
```diff
@@ -1,20 +1,86 @@
...
+  if (req_in_json.ok()) {
+    const auto http_uri = config_.http_service().http_service().http_uri();
+    const std::string cluster = http_uri.cluster();
+    const std::string uri = http_uri.uri();
+    absl::string_view host, path;
+    Envoy::Http::Utility::extractHostPathFromUri(uri, host, path);
+    ENVOY_LOG(debug, " Ext_Proc HTTP client send request to cluster {}, uri {}, host {}, path {}",
+              cluster, uri, host, path);
+
+    const auto thread_local_cluster = context().clusterManager().getThreadLocalCluster(cluster);
```
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:34]: Please add nullptr check for `thread_local_cluster ` If can be nullptr if cluster is not 
It is important  to avoid data plane crash

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc [hunk 7]
```diff
@@ -1,20 +1,86 @@
...
+             {Envoy::Http::Headers::get().RequestId, std::to_string(stream_id)},
+             {Envoy::Http::Headers::get().Host, std::string(host)}});
+    Http::RequestMessagePtr message =
+        std::make_unique<Envoy::Http::RequestMessageImpl>(std::move(headers));
+    message->body().add(req_in_json.value());
+
+    auto options = Http::AsyncClient::RequestOptions()
+                       .setTimeout(std::chrono::milliseconds(
+                           DurationUtil::durationToMilliseconds(http_uri.timeout())))
+                       .setSampled(absl::nullopt)
```
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:52]: why we want to setSampled(absl::nullopt) 
Please add : setSendInternal(false) here as well.

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc [hunk 8]
```diff
@@ -1,20 +1,86 @@
...
+    ENVOY_LOG(debug, " Ext_Proc HTTP client send request to cluster {}, uri {}, host {}, path {}",
+              cluster, uri, host, path);
+
+    const auto thread_local_cluster = context().clusterManager().getThreadLocalCluster(cluster);
+
+    // Construct a HTTP POST message and sends to the ext_proc server cluster.
+    Http::RequestHeaderMapPtr headers =
+        Envoy::Http::createHeaderMap<Envoy::Http::RequestHeaderMapImpl>(
+            {{Envoy::Http::Headers::get().Method, "POST"},
+             {Envoy::Http::Headers::get().Scheme, "http"},
```
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:40]: You can use something like 

setReferenceMethod(Http::Headers::get().MethodValues.Post);
setReferenceContentType(Http::Headers::get().ContentTypeValues.Json);

It is better than hardcode string , i don't feel strongly though

File: source/extensions/filters/http/ext_proc/http_client/client_base.h (hunk 2 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/http_client/client_base.h:25]: Yes

**Review by tyxia — COMMENTED** [review]
Thanks for working on this!

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 3 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:356]: This will incur code refactory to move the stream_  variable https://github.com/envoyproxy/envoy/blob/634f71e38b706b10d3591098dbda5eb65fc34097/source/extensions/filters/http/ext_proc/ext_proc.h#L484  from Filter object into the ExternalProcessorClientImpl object.  I would prefer not do this while we installing the basic functionality of HTTP in ext_proc.

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:356]: I added this as a TBD item in: https://github.com/envoyproxy/envoy/issues/35488

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc (hunk 8 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:39]: The key/value pair of the headers are clear to me this way.

File: source/extensions/filters/http/ext_proc/http_client/client_base.h (hunk 2 above)
Comment (yanavlasov) [reply source/extensions/filters/http/ext_proc/http_client/client_base.h:25]: When it should be moved out of the http_client directory, since this implies this is http client specific.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 3 above)
Comment (yanavlasov) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:356]: Why is it so big? Using dynamic_cast like so is an antipattern and you should not really need it. Looking at the API here you just need sendRequest that takes both end_stream and stream_id.

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:356]: The issue is for gRPC client, the client->sendRequest() eventually has to call:
stream_->send(std::move(req), end_stream);

And the stream_ is not available in class ExternalProcessorClient. Refactor it from Filter class and move it to this client will trigger some work as stream_ is is in a lot of places in the code.

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:356]: Created https://github.com/envoyproxy/envoy/pull/36228 to refactor stream_ out from Filter and moving it into client_.

**Comment by wbpcode** [comment]
Please check the list int the docs. rst's format is different with .md

image

**Comment by jmarantz** [comment]
Needs main merge. I'm having a little trouble following whether we are not blocked on reviewers or on author activity.

/wait

**Comment by yanjunxiang-google** [comment]
> Needs main merge. I'm having a little trouble following whether we are not blocked on reviewers or on author activity.
> 
> /wait

This PR is currently blocked by another PR I am working on: https://github.com/envoyproxy/envoy/pull/36228

File: source/extensions/filters/http/ext_proc/http_client/client_base.h (hunk 2 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/http_client/client_base.h]: done

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 3 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: Done of this.

I ends up passing a stream interface to the client_->sendRequest() API instead of storing the stream_ ptr in the client object.  Thus part of https://github.com/envoyproxy/envoy/pull/36228 is reworked, and stream_ ptr is put back to Filter class to make things simple.

**Comment by yanjunxiang-google** [comment]
@envoyproxy/api-shepherds PTAL and approve

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto [hunk 9]
```diff
@@ -126,11 +126,39 @@ message ExternalProcessor {
...
+  //
+  // 1. The headers are first put in a proto message
+  // :ref:`ProcessingRequest <envoy_v3_api_msg_service.ext_proc.v3.ProcessingRequest>`.
+  //
+  // 2. This proto message is then transcoded into a JSON text.
+  //
+  // 3. Envoy then sends a HTTP POST message with content-type as "application/json",
+  // and this JSON text as body to the side stream server.
+  //
+  // After receives this HTTP request message, the side stream server behavior:
```
Comment (adisuissa) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:149]:
```suggestion
  // After the side-stream receives this HTTP request message, it is expected to do as follows:
```

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto [hunk 10]
```diff
@@ -126,11 +126,39 @@ message ExternalProcessor {
...
+  //
+  // 1. It converts the body, which is a JSON string, into a ``ProcessingRequest``
+  // proto message to examine and mutate the headers.
+  //
+  // 2. It then put the mutated headers into a new proto message
```
Comment (adisuissa) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:154]:
```suggestion
  // 2. It then sets the mutated headers into a new proto message
```

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto [hunk 11]
```diff
@@ -126,11 +126,39 @@ message ExternalProcessor {
...
+  // :ref:`ProcessingResponse <envoy_v3_api_msg_service.ext_proc.v3.ProcessingResponse>`.
+  //
+  // 3. It converts this proto message ``ProcessingResponse`` into a JSON text.
```
Comment (adisuissa) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:157]:
```suggestion
  // 3. It converts ``ProcessingResponse`` proto message into a JSON text.
```

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto [hunk 12]
```diff
@@ -126,11 +126,39 @@ message ExternalProcessor {
...
+  //
+  // 4. It then sends a HTTP response back to Envoy with status code as "200",
+  // content-type as "application/json" and body as this JSON text.
```
Comment (adisuissa) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:160]:
```suggestion
  // content-type as "application/json" and sets the JSON text as the body.
```

**Review by adisuissa — COMMENTED** [review]
Thanks!
Let's make this unimplemented-hide and make it under work-in-progress so this can be tested properly and finalized later.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc (hunk 6 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:31]: done

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc (hunk 7 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc]: done

File: source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc [hunk 13]
```diff
@@ -1,20 +1,91 @@
...
+    Envoy::Http::Utility::extractHostPathFromUri(uri, host, path);
+    ENVOY_LOG(debug, " Ext_Proc HTTP client send request to cluster {}, uri {}, host {}, path {}",
+              cluster, uri, host, path);
+
+    const auto thread_local_cluster = context().clusterManager().getThreadLocalCluster(cluster);
+    if (thread_local_cluster) {
+      // Construct a HTTP POST message and sends to the ext_proc server cluster.
+      Http::RequestHeaderMapPtr headers =
+          Envoy::Http::createHeaderMap<Envoy::Http::RequestHeaderMapImpl>(
+              {{Envoy::Http::Headers::get().Method, "POST"},
```
Comment (jmarantz) [inline source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc:36]: drive-by: factor out Envoy::Http::Headers::get() into a temp. Each call has to do some sort of atomic operation or mutex for thread safety (generated by compiler) so it's wasteful to do this many times in one context.

Also it would be more readable to factor all these out.

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/http_client/http_client_impl.cc]: Good idea, done!

**Comment by yanjunxiang-google** [comment]
Kind Ping!

File: source/extensions/filters/http/ext_proc/client_impl.cc [hunk 14]
```diff
@@ -24,6 +24,13 @@ ExternalProcessorStreamPtr ExternalProcessorClientImpl::start(
                                              sidestream_watermark_callbacks);
 }
 
+void ExternalProcessorClientImpl::sendRequest(
+    envoy::service::ext_proc::v3::ProcessingRequest&& request, bool end_stream, const uint64_t,
+    RequestCallbacks*, StreamBase* stream) {
+  ExternalProcessorStream* grpc_stream = dynamic_cast<ExternalProcessorStream*>(stream);
```
Comment (yanavlasov) [inline source/extensions/filters/http/ext_proc/client_impl.cc:30]: Please avoid the use of `dynamic_cast` This is indication that interfaces are not properly designed.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/filters/http/ext_proc/client_impl.cc (hunk 14 above)
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/client_impl.cc:30]: For this dynamic_cast, passing ExternalProcessorStream* will cause circular dependency.   Thus a StreamBase* is passed in, which will be casted into ExternalProcessorStream* to send the message.

We had tried store an ExternalProcessorStream* pointer in the client_ object(https://github.com/envoyproxy/envoy/pull/36228). However, that does not work well and trigger a lot other complexities, so has to be abandoned.

So, can we take an exception here? We can discuss offline and have a follow up PR to clean this up after we find a solution.

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/client_impl.cc:30]: Added an TBD item under https://github.com/envoyproxy/envoy/issues/35488 to track this.

**Review by yanavlasov — APPROVED** [review]
